### PR TITLE
feat(contract) move initialization of space proxy to external contract

### PR DESCRIPTION
**Pull request by giuseppecrj**
During space creation, we deploy `SpaceProxy.sol` which in turn calls a set of internal functions that have been inherited by the contract, this opens up to potential security issues since an inherited contract [could have a public function](https://github.com/river-build/river/pull/161) and open itself up to hacks

This PR introduces a delegate call to a proxy initializer on space creation so the initial state can be set while removing the inheritance pattern risk.

### CODE REVIEW

**Review by pfives — APPROVED** [review]

File: contracts/src/spaces/facets/proxy/ISpaceProxyInitializer.sol [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import {IERC5643} from "contracts/src/diamond/facets/token/ERC5643/IERC5643.sol";
+import {IERC173} from "contracts/src/diamond/facets/ownable/IERC173.sol";
+
+import {IManagedProxyBase} from "contracts/src/diamond/proxy/managed/IManagedProxy.sol";
+
+// libraries
+
+// contracts
+import {ITokenOwnableBase} from "contracts/src/diamond/facets/ownable/token/ITokenOwnable.sol";
+import {IMembershipBase} from "contracts/src/spaces/facets/membership/IMembership.sol";
```
Comment (shuhuiluo) [inline contracts/src/spaces/facets/proxy/ISpaceProxyInitializer.sol:14]:
```suggestion
// interfaces
import {ITokenOwnableBase} from "contracts/src/diamond/facets/ownable/token/ITokenOwnable.sol";
import {IMembershipBase} from "contracts/src/spaces/facets/membership/IMembership.sol";
import {IManagedProxyBase} from "contracts/src/diamond/proxy/managed/IManagedProxy.sol";

// libraries

// contracts
```

File: contracts/src/spaces/facets/proxy/SpaceProxyInitializer.sol [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  function initialize(
+    address owner,
+    ManagedProxy memory managedProxy,
+    TokenOwnable memory tokenOwnable,
+    Membership memory membership
+  ) external initializer {
+    __IntrospectionBase_init();
+    __TokenOwnableBase_init(tokenOwnable);
+    __ERC721ABase_init(membership.name, membership.symbol);
+    __MembershipBase_init(membership, managedProxy.manager);
```
Comment (shuhuiluo) [inline contracts/src/spaces/facets/proxy/SpaceProxyInitializer.sol:38]: Can pass the manager address only.

File: contracts/src/factory/facets/architect/ArchitectBase.sol [hunk 3]
```diff
@@ -353,20 +358,34 @@ abstract contract ArchitectBase is Factory, IArchitectBase, PricingModulesBase {
...
-        membershipSettings
+        proxyInitializer,
+        abi.encodeCall(
+          SpaceProxyInitializer.initialize,
+          (
+            msg.sender,
+            IManagedProxyBase.ManagedProxy({
+              managerSelector: IProxyManager.getImplementation.selector,
+              manager: address(this)
+            }),
```
Comment (shuhuiluo) [inline contracts/src/factory/facets/architect/ArchitectBase.sol:381]: Should reuse the struct or pass `address(this)` only.

File: contracts/src/spaces/facets/proxy/SpaceProxyInitializer.sol (hunk 2 above)
Comment (giuseppecrj) [reply contracts/src/spaces/facets/proxy/SpaceProxyInitializer.sol]: yep, fixed in latest.

File: contracts/src/factory/facets/architect/ArchitectBase.sol (hunk 3 above)
Comment (giuseppecrj) [reply contracts/src/factory/facets/architect/ArchitectBase.sol]: fixed in latest push

File: contracts/src/spaces/facets/proxy/ISpaceProxyInitializer.sol (hunk 1 above)
Comment (giuseppecrj) [reply contracts/src/spaces/facets/proxy/ISpaceProxyInitializer.sol]: fixed in latest push.

**Review by shuhuiluo — APPROVED** [review]
Lgtm.

**Review by shuhuiluo — APPROVED** [review]

**Pull request merged**